tests/test_channels.py

```diff
@@ -1,32 +1,157 @@
+                }
+            ],
+        }
+    )
+    client.put_bucket_policy(Bucket=bucket_name, Policy=bucket_policy)
+
+    # Minio has to start with an empty directory; once available,
+    # we can import all channel files by "uploading" them
+    mamba_repo_dir = Path(__file__).parent / "data" / "mamba_repo"
+    for path in (mamba_repo_dir / "noarch").iterdir():
+        key = path.relative_to(mamba_repo_dir)
+        client.upload_file(str(path), bucket_name, str(key), ExtraArgs={"ACL": "public-read"})
+
+
+def _s3_adapter_send_boto3_patch_factory(endpoint):
+    def _send_boto3(self, boto3, resp, request):
+        """
+        We use this to patch S3Adapter._send_boto3 function so
+        it connects to our local instance. All we are changing here
+        is the call to `session.resource(...)`.
+        """
+        from botocore.exceptions import BotoCoreError, ClientError
+        from botocore.client import Config
+        from conda.common.compat import ensure_binary
+        from conda.common.url import url_to_s3_info
+        from conda.gateways.connection import CaseInsensitiveDict
+
+        bucket_name, key_string = url_to_s3_info(request.url)
+        # https://github.com/conda/conda/issues/8993
+        # creating a separate boto3 session to make this thread safe
+        session = boto3.session.Session()
+        # create a resource client using this thread's session object
+        s3 = session.resource(
+            "s3",
+            endpoint_url=endpoint,
+            aws_access_key_id="minioadmin",
+            aws_secret_access_key="minioadmin",
+            config=Config(signature_version="s3v4"),
+            region_name="us-east-1",
+        )
+        # finally get the S3 object
+        key = s3.Object(bucket_name, key_string[1:])
+
+        try:
+            response = key.get()
+        except (BotoCoreError, ClientError) as e:
+            resp.status_code = 404
+            message = {
+                "error": "error downloading file from s3",
+                "path": request.url,
+                "exception": repr(e),
+            }
+            resp.raw = self._write_tempfile(lambda x: x.write(ensure_binary(json.dumps(message))))
+            resp.close = resp.raw.close
+            return resp
+
+        key_headers = response["ResponseMetadata"]["HTTPHeaders"]
+        resp.headers = CaseInsensitiveDict(
+            {
+                "Content-Type": key_headers.get("content-type", "text/plain"),
+                "Content-Length": key_headers["content-length"],
+                "Last-Modified": key_headers["last-modified"],
+            }
+        )
+
+        resp.raw = self._write_tempfile(key.download_fileobj)
+        resp.close = resp.raw.close
+
+        return resp
+
+    return _send_boto3
+
+
+@pytest.skipif(
+    run(["minio", "-v"], check=False).returncode != 0,
+    reason="Minio server not available on PATH",
+)
+def test_s3_server(s3_server):
```

Comment:
@jezdez I am wondering whether this mocked channel tests would be useful upstream too. Most tests just use anaconda.org and that's it. We don't have S3 or FTP specific tests in `conda/conda`.

Reply:
Yes, that'd be splendid!